cpp/src/copying/pack.cpp

```diff
@@ -0,0 +1,244 @@
+  // padding explicitly so we can force initialize the value to 0, allowing us to do a byte-by-byte
+  // comparison between the output anonymized byte buffers for testing purposes.
+  // NOTE: if size_type ever grows (eg 64 bits), this field can be removed.
+  int pad;
+};
+
+/**
+ * @brief Deserialize a single column into a column_view
+ *
+ * Deserializes a single column (it's children are assumed to be already deserialized)
+ * non-recursively.
+ *
+ * @param serial_column Serialized column information
+ * @param children Children for the column
+ * @param base_ptr Base pointer for the entire contiguous buffer from which all columns
+ * were serialized into
+ * @return Fully formed column_view
+ */
+column_view deserialize_column(serialized_column serial_column,
+                               std::vector<column_view> const& children,
+                               uint8_t const* base_ptr)
+{
+  auto const data_ptr = serial_column.data_offset != -1 ? base_ptr + serial_column.data_offset : 0;
+  auto const null_mask_ptr =
+    serial_column.null_mask_offset != -1
+      ? reinterpret_cast<bitmask_type const*>(base_ptr + serial_column.null_mask_offset)
+      : 0;
+
+  return column_view(serial_column.type,
+                     serial_column.size,
+                     data_ptr,
+                     null_mask_ptr,
+                     serial_column.null_count,
+                     0,
+                     children);
+}
+
+/**
+ * @brief Build and add metadata for a column and all of it's children, recursively
+ *
+ *
+ * @param metadata Output vector of serialized_column metadata
+ * @param col Column to build metadata for
+ * @param base_ptr Base pointer for the entire contiguous buffer from which all columns
+ * were serialized into
+ * @param data_size Size of the incoming buffer
+ */
+void build_column_metadata(std::vector<serialized_column>& metadata,
+                           column_view const& col,
+                           uint8_t const* base_ptr,
+                           size_t data_size)
+{
+  uint8_t const* data_ptr = col.size() == 0 || !col.head<uint8_t>() ? nullptr : col.head<uint8_t>();
+  if (data_ptr != nullptr) {
+    CUDF_EXPECTS(data_ptr >= base_ptr && data_ptr < base_ptr + data_size,
+                 "Encountered column data outside the range of input buffer");
+  }
+  int64_t const data_offset = data_ptr ? data_ptr - base_ptr : -1;
+
+  uint8_t const* null_mask_ptr = col.size() == 0 || !col.nullable()
+                                   ? nullptr
+                                   : reinterpret_cast<uint8_t const*>(col.null_mask());
+  if (null_mask_ptr != nullptr) {
+    CUDF_EXPECTS(null_mask_ptr >= base_ptr && null_mask_ptr < base_ptr + data_size,
+                 "Encountered column null mask outside the range of input buffer");
+  }
+  int64_t const null_mask_offset = null_mask_ptr ? null_mask_ptr - base_ptr : -1;
+
+  // add metadata
+  metadata.emplace_back(serialized_column{col.type(),
```

Comment:
So this is still invoking the move ctor of `serialized_column`. You can just pass the ctor arguments directly to `serialized_column` and it will construct the object inplace. Granted, there won't be much performance difference between the move ctor and direct construction, but since `emplace_back` is already being used might as well use the more efficient thing.